feat(email-templates): refactor and add email templates from 1.48.3

Aaron-Ritter opened the pull request.

mooreds commented:
@Aaron-Ritter did you run this locally?

When I try to do so, I get this error message:

```
Failed to load url src/content/docs/_shared/email/_PasswordlessLogin-text.mdx (resolved id: src/content/docs/_shared/email/_PasswordlessLogin-text.mdx) in /path/fusionauth-site/astro/src/content/docs/customize/email-and-messages/templates-replacement-variables.mdx. Does the file exist?
Failed to load url src/content/docs/_shared/email/_BreachedPasswordNotification-text.mdx (resolved id: src/content/docs/_shared/email/_BreachedPasswordNotification-text.mdx) in /path/fusionauth-site/astro/src/content/docs/customize/email-and-messages/templates-replacement-variables.mdx. Does the file exist?
```

What am I missing?

Aaron-Ritter commented:
@mooreds I only get a not found now, which is related to the fact that it points to the fusionauth-site master branch which is not released yet, when i point it to my own branch it works.

The only part I need to know if need to convert the source files in to the lowercase-dash format too.

mooreds commented:
Hi @Aaron-Ritter please do rename the source files so they look like this:

```
breached-password.html.ftl
breached-password.txt.ftl
change-password.html.ftl
change-password.txt.ftl
confirm-child.html.ftl
confirm-child.txt.ftl
coppa-email-plus-notice.html.ftl
coppa-email-plus-notice.txt.ftl
coppa-notice.html.ftl
coppa-notice.txt.ftl
email-verification.html.ftl
email-verification.txt.ftl
parent-registration.html.ftl
parent-registration.txt.ftl
passwordless-login.html.ftl
passwordless-login.txt.ftl
registration-verification.html.ftl
registration-verification.txt.ftl
setup-password.html.ftl
setup-password.txt.ftl
threat-detected.html.ftl
threat-detected.txt.ftl
two-factor-add.html.ftl
two-factor-add.txt.ftl
two-factor-login.html.ftl
two-factor-login.txt.ftl
two-factor-remove.html.ftl
two-factor-remove.txt.ftl
```

Thanks!

Aaron-Ritter commented:
@mooreds ready for review

mooreds commented on astro/src/content/docs/customize/email-and-messages/templates-replacement-variables.mdx:
```diff
@@ -11,10 +11,10 @@ import EmailTemplateBaseUrlNote from 'src/content/docs/customize/email-and-messa
+import BreachedPasswordHTML from 'src/content/docs/_shared/email/_breached-password-notification-html.mdx';
+import BreachedPasswordText from 'src/content/docs/_shared/email/_breached-password-notification-txt.mdx';
+import PasswordlessHTML from 'src/content/docs/_shared/email/_passwordless-login-html.mdx';
+import PasswordlessText from 'src/content/docs/_shared/email/_passwordless-login-txt.mdx';
```
Please update this page to use the remote include components for all the other sample messages.

Aaron-Ritter commented:
@mooreds Replaced all email templates with the remote include throughout the full repo.

Aaron-Ritter commented:
@mooreds this is ready to be merged, as discussed i added a URL list file with one exception i did not add a comment section to the file as e.g. `wget -i <filename>` would still try and parse it.

Instead i added it in to the official documentation directly, see latest commit for this: https://github.com/FusionAuth/fusionauth-site/pull/2758/commits/828d144c0fd1f0bf8fe420947bd5c1ea1e1a4709

I tested it with my own URLs successfully and it will fully work once committed to the master.

mooreds reviewed (changes requested):
I get an error message when I run this locally.

When I visit http://localhost:3000/docs/lifecycle/authenticate-users/passwordless/magic-links

```
Cannot find module 'src/content/docs/_shared/email/_passwordless-login-text.mdx' imported from '/Users/dan/dev/inversoft/fusionauth/fusionauth-site/astro/src/content/docs/lifecycle/authenticate-users/passwordless/magic-links.mdx'
```

What am I missing?

Aaron-Ritter commented:
@mooreds `-text.mdx` endings are wrong everywhere, I changed them to `-txt.mdx` instead.

mooreds commented:
The commit you mention here: https://github.com/FusionAuth/fusionauth-site/pull/2758#issuecomment-1870351489 is not present in this PR. 

Also, I don't think we want normal users to be running that wget command. Maybe I misremember our conversation but I feel like that is something that will be done by developers updating our docs. Can you explain to me again what the purpose of that list is?

mooreds reviewed (changes requested):
See my comment about the email template list.

mooreds commented:
Okay, so now we have two means of managing email templates which is additional complexity. You can get all the default ones by using wget as you outlined in the doc and then use node-cli to move them into your system. 

Why don't we add a command to the node cli tool which pulls down all the default templates. It can still pull from this file, but then the user only needs one tool. 

We'd also need to add a note for developers to add a line to the list of urls whenever new templates are added. Can you please do that in astro/src/content/docs/customize/email-and-messages/templates-replacement-variables.mdx at the top (just a dev focused comment). 

I was incorrect about the wget -i command not being included, my apologies.

Make sense @Aaron-Ritter ?

Aaron-Ritter commented:
Correct, and actually more than two;
- During **Kickstart** with the initial setup and in the tech preview the reapplying of it.
- With **Terraform Provider** very similar to the kickstart.
- As mentioned by you the **node cli**

For Kickstart and for sure in the case of Terraform it would go again the principle of Infrastructure as Code to use a additional tooling for managing the email templates so something like `node cli` would be out of scope and a simple `wget -i` does the job of easily getting the latest templates which you then can diff with your current configuration.

Where for the `node cli` approach it would need more than a URL list. This is basically not enough information to deploy a email template as its assumed or completely missing information:
- Name (could be derived from the file name)
- From Email (could just assume its dev@fusionauth.io but it might change)
- Default from Name (could just assume its FusionAuth Development but it might change)
- Default Subject (can't be derived or assumed)

For this we would require e.g. a more complex JSON file which we then can process in the `node cli`. But for the Terraform and Kickstart scenario I would argue that it is overkill.

So my suggestion for now would be to use the `wget -i` option, add the note for developers as suggested by you and merge the PR. Create a follow-up issue to go through the different node cli scenarios and define a email template source format most suitable for it. Alternatively there could be as well a dialogue based option with the current URL list approach to fill and match what's missing, but this is something I would like to first define as different possible use cases.

mooreds commented on astro/src/content/docs/customize/email-and-messages/templates-replacement-variables.mdx:
```diff
@@ -30,6 +54,11 @@ If you create a template with content and template variables for one type of ema
 Instead, the attempt will fail and there will be an error added to the Event Log.
 </Aside>
 
```
```suggestion

### Retrieving Default Templates
```

mooreds commented on astro/src/content/docs/customize/email-and-messages/templates-replacement-variables.mdx:
```diff
@@ -30,6 +54,11 @@ If you create a template with content and template variables for one type of ema
 Instead, the attempt will fail and there will be an error added to the Event Log.
 </Aside>
 
+To get all the latest email templates at once, you can run the following command in your desired destination directory:
```
```suggestion

In order to version control the templates or customize them, you can use the admin UI. But you can also pull retrieve them all by using the following command:
```

mooreds commented on astro/src/content/docs/customize/email-and-messages/templates-replacement-variables.mdx:
````diff
@@ -30,6 +54,11 @@ If you create a template with content and template variables for one type of ema
+
+```
+wget -i https://raw.githubusercontent.com/FusionAuth/fusionauth-site/master/astro/src/content/docs/_shared/email/template_url_list
+```
````
```suggestion

This will place all the email templates in the current working directory.
```

mooreds commented on astro/src/content/docs/customize/email-and-messages/templates-replacement-variables.mdx:
```diff
@@ -11,10 +11,34 @@ import EmailTemplateBaseUrlNote from 'src/content/docs/customize/email-and-messa
+import ThreatDetectedText from 'src/content/docs/_shared/email/_threat-detected-txt.mdx';
+import TwoFactorAuthenticationMethodAddedHTML from 'src/content/docs/_shared/email/_two-factor-authentication-method-added-html.mdx';
+import TwoFactorAuthenticationMethodAddedText from 'src/content/docs/_shared/email/_two-factor-authentication-method-added-txt.mdx';
 
```
```suggestion

[//]: # "MAKE SURE YOU UPDATE astro/src/content/docs/_shared/email/template_url_list if you add any templates"

```

mooreds reviewed (changes requested):
I get your point about `wget -i`. Makes sense. 

A few more suggested word changes and then I think this is ready to go.
